File: experiment.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

"""
Perform bayesian variable selection to obtain inclusion probabilities.
"""

import argparse
import logging

import pandas as pd

from src.common.df_csv_writing import write_df_to_csv
from src.models.bayesian_model_averaging.model import ODAPerformer

# xgboost
from xgboost import XGBClassifier
from xgboost import plot_importance
from matplotlib import pyplot

logger = logging.getLogger(__name__)

if __name__ == "__main__":

    # configure parser and parse arguments
    parser = argparse.ArgumentParser(description='Perform bayesian variable selection on dataset.')
    parser.add_argument('--dataset', type=str, help='The path to the dataset file', required=True)
    parser.add_argument('--csv_separator', type=str, help='The separator of the data columns', default=',')
    parser.add_argument('--importance_method', type=str, help='The method to determine feature importance. Options: bma=complete dataset, xgb=incomplete dataset', default='xgb')
    parser.add_argument('--niter', type=int, default=100000, help='The number of iterations to perform')
    parser.add_argument('--burn_in_sim', type=int, default=500, help='Burn in sim')
    parser.add_argument('--lam_spec', type=float, default=1, help='lam spec')
    parser.add_argument('--label', type=str, default='diagnostic_outcome', help='The name of the label column.')
    parser.add_argument('--show_plots', type=bool, help='Show plots if they are available', default=False)

    args = parser.parse_args()
    dataset_path = args.dataset
    csv_separator = args.csv_separator
    importance_method = args.importance_method
    iteration_qty = args.niter
    burn_in_sim = args.burn_in_sim
    lam_spec = args.lam_spec
    label = args.label
    show_plots = args.show_plots

    logger.info(str({"message": "NEW EXPERIMENT",
                     "path": dataset_path,
                     "iteration_qty": iteration_qty,
                     "burn_in_sim": burn_in_sim,
                     "lam_spec": lam_spec})
                )

    logger.info(str({"message": "Load dataset",
                     "path": dataset_path}))
    mi_df = pd.read_csv(dataset_path, header=0, sep=csv_separator)  # read data from csv

    logger.info("Sanity check:")
    logger.info(mi_df.shape)
    logger.info(mi_df.columns)

    if mi_df.isnull().values.any() and importance_method == 'bma':
        logger.critical(str({"messsage": "BMA can only work with a complete/imputed dataset. "
                                         "Run imputation first using dataset.py or dataset.r"}))
        exit(-1)

    # prepare data for bayesian variable selection
    feature_names = [column_name for column_name in list(mi_df) if column_name != label]  # extract feature names
    features = mi_df[feature_names]  # extract features
    labels = mi_df[label]  # extract label

    incprobs_df = pd.DataFrame(columns=feature_names)

    if importance_method == "bma":
        # Calculate oda inclusion probabilities for ICD-10
        oda_results = ODAPerformer(iteration_qty=iteration_qty, burn_in_sim=burn_in_sim, lam_spec=lam_spec).fit_transform(features, labels)

        # create a new dataset for inclusion probabilities
        incprobs_df.loc[0] = oda_results["incprob_rb"]

        logger.info(str({"message": "Results",
                         "inclusion_probabilities": incprobs_df})
                    )

        # logger.info(oda_results["betama"])
        # logger.info(oda_results["incprob"])
        # logger.info(oda_results["gamma"])
        # logger.info(oda_results["odds"])

    elif importance_method == "xgb":
        # xgradient boosting
        # source: https://machinelearningmastery.com/feature-importance-and-feature-selection-with-xgboost-in-python/
        # handling of missing values: https://stats.stackexchange.com/questions/235489/xgboost-can-handle-missing-data-in-the-forecasting-phase
        # https://stackoverflow.com/questions/37617390/xgboost-handling-of-missing-values-for-split-candidate-search

        # fit model to training data
        model = XGBClassifier()
        model.fit(features, labels)

        if show_plots:
            # plot feature importance
            plot_importance(model)
            pyplot.show()

        # create a new dataset for feature importances
        incprobs_df.loc[0] = model.feature_importances_

        logger.info(str({"message": "Results",
                         "feature_importances": incprobs_df})
                    )

        # TODO: Setup a train/test setting to validate your model

    # write dataset to file
    write_df_to_csv(df=incprobs_df, store_path='data/processed/', initial_path=dataset_path, file_appendix='_importance_method_' + importance_method + '_incprobs')

File: experiment.r
###
# Perform bayesian variable selection to obtain inclusion probabilities.
###

library(optparse) # parse script arguments in a pythonic way
source("./src/models/bayesian_model_averaging/oda_bma/calc-modelprobs.r", chdir=TRUE)
source("./src/models/bayesian_model_averaging/oda_bma/oda.bma.r", chdir=TRUE)

# configure parser and parse arguments
optionList = list(
make_option(c("--dataset"),
type="character", default="./data/interim/20140721000003_myocardial_ischemia_16_impType_MI_nIter_20_label.csv", help="Path to the dataset file", metavar="character"),
make_option(c("--label"),
type="character", default="diagnostic_outcome", help="Label of the dataset", metavar="character"),
make_option(c("--niter"),
type="integer", default=1000, help="The number of iterations to perform", metavar = "integer"),
make_option(c("--burnIn"),
type="integer", default=500, help="The number of iterations we discard to tune the initial probabilities (burn-in)", metavar = "integer"),
make_option(c("--lambda"),
type="double", default=1, help="lambda", metavar = "double"),
make_option(c("--coeffShrink"),
type="double", default=0, help="regularization through coefficient shrinking", metavar = "double"),
make_option(c("--ridgeLassoBlend"),
type="double", default=0, help="Blend L2 regularization (ridge) with L1 regularization (LASSO). regularization = (1-ridgeLassoBlend) * ridge + ridgeLassoBlend * LASSO ", metavar = "double"),
make_option(c("--appendix"),
type="character", help="Optional repetition id of the experiment to explore convergence across multiple chains", metavar = "character"))

# parse script arguments
optParser <- OptionParser(option_list=optionList)
opt <- parse_args(optParser)

if (is.null(opt$dataset)){  # print and stop script if dataset file path is missing
    print_help(optParser)
    stop("Missing dataset file argument")
}
datasetPath <- opt$dataset
iterationQty <- opt$niter
burnIn <- opt$burnIn
lambda <- opt$lambda
coeffShrink <- opt$coeffShrink
ridgeLassoBlend <- opt$ridgeLassoBlend
label <- opt$label
if (is.null(opt$appendix)){
    appendix <- ""
} else {
    appendix <- paste0("_", opt$appendix)
}

# load data from csv
cat("Loading data from csv...")
miData<-read.csv(opt$dataset, sep=",", header=TRUE)
cat("...Done.\n")

# Calculate oda inclusion probabilities for ICD-10
cat("Sanity check for dataset:\n") # check if dimensions and names of matrix look sane
print(dim(miData)) # 3424 15
print(names(miData))
print("---------------------")

# extract feature names
columnNames <- colnames(miData)
isNotLabel <- function(x){x != label}
featureNames <- Filter(isNotLabel, columnNames)

print("Features:")
print(featureNames)
print("Label:")
print(label)

features <- subset(miData, select=featureNames)  # extract features
labels <- miData[label]  # extract label

print(paste0("Sampling models using oda-bma method to estimate inclusion probabilities of features for ", iterationQty, " iterations..."))
odaResults <- oda.bma(x = features, y = labels, niter = iterationQty, burnin = burnIn, lambda = lambda, coeffShrink=coeffShrink, ridgeLassoBlend=ridgeLassoBlend, model = "probit", prior = "normal")

# print(odaResults$incprob.rb)
# print(odaResults$betabma)
# print(odaResults$incprob)
# print(odaResults$gamma)
# print(odaResults$odds)
print("...Done.\n")

print("Calculating posterior probability of unsampled models...")
# matrix of unique models from ODA
gamma.u <- unique(odaResults$gamma[-c(1:burnIn),])

# vector of RB estimates of model probabilities corresponding to models in gamma.u
probest <- modelprobs.rb(n.unique=nrow(gamma.u),
						niter=nrow(odaResults$incprob[-c(1:burnIn),]),
						p = ncol(gamma.u),
						gammaunique=t(gamma.u),
						probmat.oda = t(odaResults$incprob[-c(1:burnIn),]))
print("...Done.\n")


print("Results:\n")
print("----------")

print("Estimated posterior probability of unsampled models (= the size of the unsampled model space, should be < 0.05, 0.025 or 0.01):")
print(1 - sum(probest))

# create a new dataset for inclusion probabilities
incprobsDf <- data.frame(matrix(ncol = length(featureNames), nrow = 0))
colnames(incprobsDf) <- featureNames
incprobsDf[1,] <- odaResults$incprob.rb

print("Inclusion probabilities of features:")
print(incprobsDf)

print("----------")

cat("Writing dataset to file...")
fileName <- sub(pattern = "(.*?)\\.[a-zA-Z]*$", replacement = "\\1", basename(datasetPath))

# prepare dataset store path
path <- 'data/processed/'
if(!grepl("[0-9]{14}", fileName)){  # try to find a timestamp with 4 digit year and each 2 digits for month, day, hour, minute, second
    # write imputed data to file with timestamp
    cat("Writing imputed data to file...")
    now <- Sys.time()
    path <- paste0(path, format(now, "%Y%m%d%H%M%S"), "_")
}
path <- paste0(path, fileName, "_niter_", iterationQty, "_burnIn_", burnIn, "_ridgeLasso_", ridgeLassoBlend, "_shrink_", coeffShrink, "_incprobs", appendix, ".csv")
print(path)
write.csv(incprobsDf, file=path, row.names = FALSE)
cat("...Done.")
